### ISSUE TITLE
Entry points on Windows do not set __package__ correctly

### ISSUE THREAD

**Issue by davidism** [pypa/setuptools#1838]
Werkzeug provides a reloader for WSGI applications like Flask. Because Python behaves slightly differently when run as `python file` vs `python -m module`, the reloader needs to figure out how Python was invoked in order to invoke it the same way on reload. To determine if `-m module` was used, it looks at `__main__.__package__`. If it's set to `None`, it was run as a file, otherwise it was run as a package. See explanation [here](https://stackoverflow.com/q/21233229/400617) and in [PEP 366](https://www.python.org/dev/peps/pep-0366/#proposed-change):

> When the main module is specified by its filename, then the `__package__` attribute will be set to `None`.

However, entry points on Windows are not behaving this way. Instead, `__package__` is the empty string when running an entry point. This caused an issue with the Werkzeug 0.15.5 release, see pallets/werkzeug#1614.

Additionally, entry points on Windows *are* behaving correctly when the package is installed in editable mode (`pip install -e .`), and they behave consistently and correctly on other platforms. So there seems to be an inconsistency in how setuptools creates entry points on Windows.

The following two files reproduce this. Use `pip install -e .` then run `ep-example`, observe the output. Then use `pip install .` and `ex-example`, and observe that the output is different. You can confirm that the output is consistent by doing the same on Linux.

`ep_example.py`:

```python
import sys

def main():
    __main__ = sys.modules["__main__"]
    print("__main__.__package__:", repr(__main__.__package__))
```

`setup.py`:

```python
from setuptools import setup

setup(
    name="ep-example",
    version="0.1.0",
    py_modules=["ep_example"],
    entry_points={"console_scripts": ["ep-example=ep_example:main"]},
)
```

Correct behavior for editable install on Windows, and editable and dist installs on Linux or Mac:

```
> pip install -e .
> ep-example
__main__.__package__: None
```

Incorrect behavior for dist install on Windows:

```
> pip install .
> ep-example
__main__.__package__: ''
```

**Comment by pganssle** [comment]
I think this might be a bug in `pip`, not in `setuptools`. The editable install behavior is using `setuptools` to do the *installation* and so `setuptools` (or possibly `easy_install`) is generating the entry points, which is why it uses `pkg_resources`:


This is the script generated by `pip install -e .`:

```python
#!/tmp/wzep/venv/bin/python
# EASY-INSTALL-ENTRY-SCRIPT: 'ep-example','console_scripts','ep-example'
__requires__ = 'ep-example'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('ep-example', 'console_scripts', 'ep-example')()
    )
```

This is the version installed by `pip`:

```python
#!/tmp/wzep/venv/bin/python
# -*- coding: utf-8 -*-
import re
import sys

from ep_example import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

@pypa/pip-helpers @pypa/pip-committers Should we move this over to `pip`'s issue tracker?

**Comment by davidism** [comment]
`pip` does generate cleaner entry points, but I think I observed the same behavior with `python setup.py install` vs `develop`, let me try again.

**Comment by davidism** [comment]
Hmm, unless I've messed up something, it looks like `setup.py install` creates an entry point that does the right thing, while `pip install .` does not. So this might need to be moved to pip.

**Comment by chrahunt** [comment]
The actual change may end up being in [pypa/distlib](https://bitbucket.org/pypa/distlib), which pip vendors.

**Comment by davidism** [comment]
Is there anything you'd like me to do to help? I tried to look at how entry points were generated and had a hard time following it. Should this be moved to a different repo?

**Comment by pganssle** [comment]
CC @merwok, who appears to be the maintainer for `distlib`. Given that `distlib` is on bitbucket (for now, at least given that BB is dropping mercurial support), I don't think we can just move the issue there, maybe open a new issue on that repo?

**Comment by merwok** [comment]
Sorry, I am not a developer of setuptools, distlib or pip.

Vinay Sajip is the author of distlib.

**Comment by pganssle** [comment]
> Sorry, I am not a developer of setuptools, distlib or pip.
> 
> Vinay Sajip is the author of dis

Oh my mistake, I misread "[CONTRIBUTORS.txt](https://bitbucket.org/pypa/distlib/src/default/CONTRIBUTORS.txt)" from the repo as saying you were the maintainer, but that actually seems to be taken wholesale from `distlib2`.  CC @vsajip

**Comment by chrahunt** [comment]
For completeness - currently in pip we use a custom entrypoint script text but that will be removed in pypa/pip#6763 and ideally we would not have to reintroduce it i.e. we can incorporate this into distlib directly.

**Comment by vsajip** [comment]
Even with usage of the new script template from `distlib`, I'm not sure that will fix this issue - AFAIK `distlib` is only used to generate the launcher executables.

**Comment by davidism** [comment]
I'm not clear that it's the script, since the pip and distlib scripts behave correctly on Linux. Maybe it's some quirk of the exe.

**Comment by vsajip** [comment]
> Maybe it's some quirk of the exe.

Could be, but it seems unlikely - the exe's job is just to invoke the correct Python interpreter, passing it the exe as an argument (along with other arguments passed to it).

**Comment by pfmoore** [comment]
I wonder, could the difference be running the script from what is in effect a zipfile, rather than as a standalone script? I'm not sufficiently clear on how to reproduce the issue without an entry point (i.e., how to demonstrate "`__package__` is set to X" using just a Python script) to check that hypothesis myself.

**Comment by davidism** [comment]
As part of pallets/werkzeug#1416, which was the original work on Werkzeug's reloader to understand `python -m module`, I did install an egg with an entry point, and that seemed to work on Linux. I used `easy_install` so it wasn't unzipped.
